File: ios-native/PearBrowser/Sources/UI/Screens/SiteEditorScreen.swift
//  PearBrowser — SiteEditorScreen.swift
//
//  SwiftUI mirror of app/screens/SiteEditorScreen.tsx — simplified first
//  pass. Supports the 8 block types, add/remove/reorder, save, publish,
//  preview. Theme picker is a 5-preset row. Published URL opens in the
//  Browse tab.
//
//  Backend wire calls CMD_UPDATE_SITE + CMD_PUBLISH_SITE.

import SwiftUI

enum BlockType: String, CaseIterable {
    case heading, text, list, divider, code, quote, link, image

    var label: String {
        switch self {
        case .heading: return "H"
        case .text: return "T"
        case .list: return "="
        case .divider: return "--"
        case .code: return "{}"
        case .quote: return "\""
        case .link: return "@"
        case .image: return "Img"
        }
    }
}

struct EditorBlock: Identifiable, Hashable {
    var id: String = UUID().uuidString
    var type: BlockType
    var text: String = ""
    var level: Int = 2      // for heading
    var items: [String] = []  // for list
    var href: String = ""    // for link
    var src: String = ""     // for image
    var alt: String = ""     // for image

    func toDict() -> [String: Any] {
        var d: [String: Any] = ["id": id, "type": type.rawValue]
        switch type {
        case .heading: d["text"] = text; d["level"] = level
        case .text, .code, .quote: d["text"] = text
        case .list: d["items"] = items
        case .link: d["text"] = text; d["href"] = href
        case .image: d["src"] = src; d["alt"] = alt
        case .divider: break
        }
        return d
    }

    static func from(_ dict: [String: Any]) -> EditorBlock? {
        guard let typeRaw = dict["type"] as? String, let type = BlockType(rawValue: typeRaw) else { return nil }
        var b = EditorBlock(id: (dict["id"] as? String) ?? UUID().uuidString, type: type)
        b.text = (dict["text"] as? String) ?? ""
        b.level = (dict["level"] as? Int) ?? 2
        b.items = (dict["items"] as? [String]) ?? []
        b.href = (dict["href"] as? String) ?? ""
        b.src = (dict["src"] as? String) ?? ""
        b.alt = (dict["alt"] as? String) ?? ""
        return b
    }
}

struct SiteEditorScreen: View {
    let siteId: String
    let siteName: String?
    /// Initial block list (from TemplatePicker) or nil to load from backend.
    let initialBlocks: [[String: Any]]?
    let initialTheme: [String: String]?
    let onBack: () -> Void
    let onPreview: (String) -> Void

    @Environment(\.pearRPC) private var rpc

    @State private var blocks: [EditorBlock] = []
    @State private var themeName: String = "default"
    @State private var saving = false
    @State private var publishing = false
    @State private var errorMessage: String? = nil
    @State private var publishResult: String? = nil

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(siteName ?? "Editor", onBack: onBack) {
                HStack(spacing: 12) {
                    Button(action: save) {
                        Text(saving ? "…" : "Save")
                            .font(.system(size: 14, weight: .semibold))
                            .foregroundStyle(PearColors.accent)
                    }
                    .disabled(saving || rpc == nil)
                    Button(action: publish) {
                        Text(publishing ? "…" : "Publish")
                            .font(.system(size: 14, weight: .semibold))
                            .foregroundStyle(PearColors.accent)
                    }
                    .disabled(publishing || rpc == nil)
                }
            }

            if let errorMessage {
                Text(errorMessage)
                    .font(.system(size: 13))
                    .foregroundStyle(PearColors.error)
                    .padding(12)
                    .frame(maxWidth: .infinity)
                    .background(Color(red: 0.17, green: 0.05, blue: 0.05))
            }

            ScrollView {
                VStack(spacing: 10) {
                    themeRow
                    ForEach($blocks) { $block in
                        blockEditor(for: $block)
                    }
                }
                .padding(16)
                .padding(.bottom, 140)
            }

            toolbar
        }
        .background(PearColors.bg)
        .onAppear { loadIfNeeded() }
        .alert("Site published", isPresented: Binding(get: { publishResult != nil }, set: { if !$0 { publishResult = nil } })) {
            Button("View") {
                if let publishResult { onPreview(publishResult) }
            }
            Button("OK", role: .cancel) {}
        } message: {
            if let publishResult { Text("Live at \(publishResult)") }
        }
    }

    // MARK: - Theme row

    private var themeRow: some View {
        let themes: [(id: String, name: String, color: Color)] = [
            ("default", "Default", PearColors.accent),
            ("dark", "Dark", Color.black),
            ("warm", "Warm", Color(red: 0.95, green: 0.55, blue: 0.2)),
            ("ocean", "Ocean", PearColors.link),
            ("forest", "Forest", PearColors.success),
        ]
        return ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(themes, id: \.id) { t in
                    Button { themeName = t.id } label: {
                        HStack(spacing: 6) {
                            Circle().fill(t.color).frame(width: 10, height: 10)
                            Text(t.name).font(.system(size: 12, weight: .medium))
                                .foregroundStyle(themeName == t.id ? PearColors.bg : PearColors.textSecondary)
                        }
                        .padding(.horizontal, 12).padding(.vertical, 6)
                        .background(themeName == t.id ? PearColors.accent : PearColors.surface,
                                    in: Capsule())
                    }
                    .buttonStyle(.plain)
                }
            }
        }
    }

    // MARK: - Block editor

    @ViewBuilder
    private func blockEditor(for block: Binding<EditorBlock>) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack {
                Text(block.wrappedValue.type.rawValue.uppercased())
                    .font(.system(size: 10, weight: .semibold))
                    .foregroundStyle(PearColors.textMuted)
                    .tracking(1)
                Spacer()
                Button { move(block.wrappedValue.id, by: -1) } label: {
                    Text("↑").font(.system(size: 14, weight: .bold)).foregroundStyle(PearColors.textSecondary)
                }
                .buttonStyle(.plain)
                Button { move(block.wrappedValue.id, by: 1) } label: {
                    Text("↓").font(.system(size: 14, weight: .bold)).foregroundStyle(PearColors.textSecondary)
                }
                .buttonStyle(.plain)
                Button { remove(block.wrappedValue.id) } label: {
                    Text("x").font(.system(size: 14, weight: .bold)).foregroundStyle(PearColors.error)
                }
                .buttonStyle(.plain)
            }
            switch block.wrappedValue.type {
            case .heading:
                TextField("Heading", text: block.text)
                    .font(.system(size: 20, weight: .bold))
                    .foregroundStyle(PearColors.textPrimary)
                    .padding(10)
                    .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 8))
            case .text, .code, .quote:
                TextEditor(text: block.text)
                    .font(block.wrappedValue.type == .code
                          ? .system(size: 13, design: .monospaced)
                          : .system(size: 14))
                    .foregroundStyle(PearColors.textPrimary)
                    .scrollContentBackground(.hidden)
                    .padding(8)
                    .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 8))
                    .frame(minHeight: 80)
            case .list:
                ForEach(Array(block.items.wrappedValue.enumerated()), id: \.offset) { idx, _ in
                    TextField("Item", text: Binding(
                        get: { block.wrappedValue.items[idx] },
                        set: { block.wrappedValue.items[idx] = $0 }
                    ))
                    .font(.system(size: 14))
                    .foregroundStyle(PearColors.textPrimary)
                    .padding(10)
                    .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 8))
                }
                Button("+ item") { block.wrappedValue.items.append("") }
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(PearColors.accent)
            case .divider:
                Rectangle().fill(PearColors.border).frame(height: 1)
                    .padding(.vertical, 8)
            case .link:
                TextField("Label", text: block.text)
                    .font(.system(size: 14))
                    .foregroundStyle(PearColors.textPrimary)
                    .padding(10)
                    .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 8))
                TextField("https://…", text: block.href)
                    .font(.system(size: 13, design: .monospaced))
                    .foregroundStyle(PearColors.link)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .padding(10)
                    .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 8))
            case .image:
                TextField("Image URL", text: block.src)
                    .font(.system(size: 13, design: .monospaced))
                    .foregroundStyle(PearColors.textPrimary)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .padding(10)
                    .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 8))
                TextField("Alt text", text: block.alt)
                    .font(.system(size: 13))
                    .foregroundStyle(PearColors.textSecondary)
                    .padding(10)
                    .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 8))
            }
        }
        .padding(12)
        .background(PearColors.surface.opacity(0.5), in: RoundedRectangle(cornerRadius: 10))
    }

    // MARK: - Toolbar

    private var toolbar: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 6) {
                ForEach(BlockType.allCases, id: \.self) { type in
                    Button { add(type) } label: {
                        Text(type.label)
                            .font(.system(size: 14, weight: .bold, design: .monospaced))
                            .foregroundStyle(PearColors.textPrimary)
                            .padding(.horizontal, 12).padding(.vertical, 8)
                            .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 8))
                    }
                    .buttonStyle(.plain)
                }
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 10)
        }
        .background(PearColors.surface.opacity(0.9))
        .overlay(alignment: .top) {
            Rectangle().fill(PearColors.border).frame(height: 0.5)
        }
    }

    // MARK: - Actions

    private func loadIfNeeded() {
        guard blocks.isEmpty else { return }
        if let initialBlocks {
            blocks = initialBlocks.compactMap { EditorBlock.from($0) }
            if blocks.isEmpty { seedDefault() }
        } else {
            seedDefault()
        }
    }

    private func seedDefault() {
        blocks = [
            EditorBlock(type: .heading, text: siteName ?? "My Website", level: 1),
            EditorBlock(type: .text, text: "Welcome to my P2P website."),
        ]
    }

    private func add(_ type: BlockType) {
        var block = EditorBlock(type: type)
        if type == .heading { block.text = "Heading" }
        if type == .text { block.text = "" }
        if type == .list { block.items = ["Item 1"] }
        if type == .link { block.text = "Link"; block.href = "https://" }
        if type == .image { block.alt = "" }
        blocks.append(block)
    }

    private func remove(_ id: String) {
        blocks.removeAll { $0.id == id }
    }

    private func move(_ id: String, by delta: Int) {
        guard let idx = blocks.firstIndex(where: { $0.id == id }) else { return }
        let target = idx + delta
        guard target >= 0 && target < blocks.count else { return }
        blocks.swapAt(idx, target)
    }

    private func save() {
        guard let rpc else { return }
        saving = true
        errorMessage = nil
        Task {
            do {
                _ = try await rpc.request(Cmd.UPDATE_SITE, data: [
                    "siteId": siteId,
                    "blocks": blocks.map { $0.toDict() },
                    "theme": themeName,
                ])
            } catch {
                errorMessage = "Save failed: \(error.localizedDescription)"
            }
            saving = false
        }
    }

    private func publish() {
        guard let rpc else { return }
        publishing = true
        errorMessage = nil
        Task {
            do {
                // Save current blocks first
                _ = try await rpc.request(Cmd.UPDATE_SITE, data: [
                    "siteId": siteId,
                    "blocks": blocks.map { $0.toDict() },
                    "theme": themeName,
                ])
                let resp = try await rpc.request(Cmd.PUBLISH_SITE, data: ["siteId": siteId])
                if let obj = resp as? [String: Any], let url = obj["url"] as? String {
                    publishResult = url
                } else if let obj = resp as? [String: Any], let key = obj["keyHex"] as? String {
                    publishResult = "hyper://\(key)"
                }
            } catch {
                errorMessage = "Publish failed: \(error.localizedDescription)"
            }
            publishing = false
        }
    }
}

File: android-native/app/src/main/java/com/pearbrowser/app/ui/screens/SiteEditorScreen.kt
package com.pearbrowser.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.pearbrowser.app.rpc.LocalPearRpc
import com.pearbrowser.app.ui.theme.PearColors
import java.util.UUID
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray

/**
 * SiteEditorScreen — mirror of ios-native `SiteEditorScreen.swift`.
 *
 * The block model mirrors backend/site-manager.js `_renderBlocks` exactly —
 * the 8 types below are all the backend understands. Save goes through
 * CMD_UPDATE_SITE with `{ siteId, blocks, theme }`, which the backend routes
 * to `siteManager.buildFromBlocks` (see backend/index.js). The theme object
 * feeds `_renderThemeCss` (primaryColor / backgroundColor / textColor /
 * fontFamily).
 *
 * There is no GET_SITE_BLOCKS RPC — opening an existing site starts from a
 * fresh default outline and saving replaces the rendered homepage, exactly
 * like the iOS editor.
 */
enum class BlockType(val wire: String, val toolbarLabel: String) {
    Heading("heading", "H"),
    Text("text", "T"),
    List("list", "="),
    Divider("divider", "--"),
    Code("code", "{}"),
    Quote("quote", "\""),
    Link("link", "@"),
    Image("image", "Img"),
}

/** One editable block. Wire shape is `toJson()` — backend/site-manager.js. */
data class EditorBlock(
    val id: String = UUID.randomUUID().toString(),
    val type: BlockType,
    val text: String = "",
    val level: Int = 2,
    val items: List<String> = emptyList(),
    val href: String = "",
    val src: String = "",
    val alt: String = "",
) {
    fun toJson(): JsonObject = buildJsonObject {
        put("type", type.wire)
        when (type) {
            BlockType.Heading -> {
                put("text", text)
                put("level", level)
            }
            BlockType.Text, BlockType.Code, BlockType.Quote -> put("text", text)
            BlockType.List -> putJsonArray("items") { items.forEach { add(it) } }
            BlockType.Link -> {
                put("text", text)
                put("href", href)
            }
            BlockType.Image -> {
                put("src", src)
                put("alt", alt)
            }
            BlockType.Divider -> Unit
        }
    }

    /** Fresh copy with a new id — used when instantiating template blocks. */
    fun instantiate(): EditorBlock = copy(id = UUID.randomUUID().toString())
}

/** Theme preset; keys feed backend/site-manager.js `_renderThemeCss`. */
data class SiteTheme(
    val id: String,
    val label: String,
    val primaryColor: String,
    val backgroundColor: String = "#0a0a0a",
    val textColor: String = "#e0e0e0",
    val fontFamily: String = "-apple-system, sans-serif",
) {
    fun toJson(): JsonObject = buildJsonObject {
        put("primaryColor", primaryColor)
        put("backgroundColor", backgroundColor)
        put("textColor", textColor)
        put("fontFamily", fontFamily)
    }

    companion object {
        // Preset row mirrors iOS SiteEditorScreen's themes; swatch colors are
        // the Pear palette values for the same names.
        val presets = listOf(
            SiteTheme("default", "Default", "#ff9500"),
            SiteTheme("dark", "Dark", "#e0e0e0"),
            SiteTheme("warm", "Warm", "#f28c33"),
            SiteTheme("ocean", "Ocean", "#4dabf7"),
            SiteTheme("forest", "Forest", "#4ade80"),
        )
        val default = presets.first()
        fun byId(id: String?): SiteTheme = presets.firstOrNull { it.id == id } ?: default
    }
}

@Composable
fun SiteEditorScreen(
    siteId: String,
    siteName: String?,
    published: Boolean,
    templateId: String?,
    onBack: () -> Unit,
    onPreview: (String) -> Unit,
) {
    val rpc = LocalPearRpc.current
    val scope = rememberCoroutineScope()
    val template = remember(templateId) { templateId?.let(SiteTemplates::byId) }

    val blocks = remember {
        mutableStateListOf<EditorBlock>().apply {
            val initial = template?.blocks.orEmpty()
            if (initial.isNotEmpty()) {
                addAll(initial.map { it.instantiate() })
            } else {
                // Default seed, mirrors iOS `seedDefault()`.
                add(EditorBlock(type = BlockType.Heading, text = siteName ?: "My Website", level = 1))
                add(EditorBlock(type = BlockType.Text, text = "Welcome to my P2P website."))
            }
        }
    }
    var theme by remember { mutableStateOf(template?.theme ?: SiteTheme.default) }
    var isPublished by remember { mutableStateOf(published) }
    var saving by remember { mutableStateOf(false) }
    var publishing by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var publishResult by remember { mutableStateOf<String?>(null) }

    suspend fun saveBlocks() {
        val client = rpc ?: throw IllegalStateException("P2P engine is not connected yet")
        client.updateSite(
            siteId = siteId,
            blocks = JsonArray(blocks.map { it.toJson() }),
            theme = theme.toJson(),
        )
    }

    fun save() {
        if (saving) return
        saving = true
        errorMessage = null
        scope.launch {
            try {
                saveBlocks()
            } catch (e: Throwable) {
                errorMessage = "Save failed: ${e.message ?: "unknown error"}"
            } finally {
                saving = false
            }
        }
    }

    fun publish() {
        if (publishing) return
        publishing = true
        errorMessage = null
        scope.launch {
            try {
                saveBlocks()
                val client = rpc ?: throw IllegalStateException("P2P engine is not connected yet")
                val resp = client.publishSite(siteId)
                publishResult = resp["url"]?.jsonPrimitive?.contentOrNull
                    ?: resp["keyHex"]?.jsonPrimitive?.contentOrNull?.let { "hyper://$it" }
                isPublished = true
            } catch (e: Throwable) {
                errorMessage = "Publish failed: ${e.message ?: "unknown error"}"
            } finally {
                publishing = false
            }
        }
    }

    fun move(index: Int, delta: Int) {
        val target = index + delta
        if (target < 0 || target >= blocks.size) return
        val tmp = blocks[index]
        blocks[index] = blocks[target]
        blocks[target] = tmp
    }

    Column(
        Modifier
            .fillMaxSize()
            .background(PearColors.Bg),
    ) {
        // Header: back, title + publish badge, Save / Publish (iOS ScreenHeader).
        Row(
            Modifier
                .fillMaxWidth()
                .background(PearColors.Bg)
                .padding(horizontal = 10.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            TextButton(onClick = onBack) {
                Text("< Back", color = PearColors.Accent, fontSize = 14.sp)
            }
            Column(Modifier.weight(1f)) {
                Text(
                    siteName ?: "Editor",
                    color = PearColors.TextPrimary,
                    fontSize = 20.sp,
                    fontWeight = FontWeight.Bold,
                    maxLines = 1,
                )
                PublishBadge(isPublished)
            }
            Text(
                if (saving) "…" else "Save",
                color = if (rpc == null || saving) PearColors.TextMuted else PearColors.Accent,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier
                    .clickable(enabled = rpc != null && !saving) { save() }
                    .padding(horizontal = 8.dp, vertical = 6.dp),
            )
            Text(
                if (publishing) "…" else "Publish",
                color = if (rpc == null || publishing) PearColors.TextMuted else PearColors.Accent,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier
                    .clickable(enabled = rpc != null && !publishing) { publish() }
                    .padding(horizontal = 8.dp, vertical = 6.dp),
            )
        }

        errorMessage?.let {
            Text(
                it,
                color = PearColors.Error,
                fontSize = 13.sp,
                modifier = Modifier
                    .fillMaxWidth()
                    .background(Color(0xFF2B0D0D))
                    .padding(12.dp),
            )
        }

        Column(
            Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp),
        ) {
            Spacer(Modifier.height(8.dp))
            ThemeRow(selected = theme, onSelect = { theme = it })
            Spacer(Modifier.height(10.dp))
            blocks.forEachIndexed { index, block ->
                BlockEditorCard(
                    block = block,
                    onChange = { blocks[index] = it },
                    onMove = { delta -> move(index, delta) },
                    onRemove = { blocks.removeAt(index) },
                )
                Spacer(Modifier.height(10.dp))
            }
            if (blocks.isEmpty()) {
                Text(
                    "No blocks yet — add one below.",
                    color = PearColors.TextMuted,
                    fontSize = 13.sp,
                    modifier = Modifier.padding(vertical = 16.dp),
                )
            }
            Spacer(Modifier.height(120.dp))
        }

        BlockToolbar(onAdd = { type -> blocks.add(newBlock(type)) })
    }

    publishResult?.let { url ->
        AlertDialog(
            onDismissRequest = { publishResult = null },
            containerColor = PearColors.Surface,
            title = { Text("Site published", color = PearColors.TextPrimary, fontWeight = FontWeight.Bold) },
            text = { Text("Live at $url", color = PearColors.TextSecondary, fontSize = 14.sp) },
            confirmButton = {
                TextButton(onClick = {
                    publishResult = null
                    onPreview(url)
                }) {
                    Text("View", color = PearColors.Accent, fontWeight = FontWeight.Bold)
                }
            },
            dismissButton = {
                TextButton(onClick = { publishResult = null }) {
                    Text("OK", color = PearColors.TextSecondary)
                }
            },
        )
    }
}

@Composable
private fun PublishBadge(published: Boolean) {
    Text(
        if (published) "Live" else "Draft",
        color = if (published) PearColors.Success else PearColors.TextMuted,
        fontSize = 11.sp,
        fontWeight = FontWeight.SemiBold,
        modifier = Modifier
            .background(
                if (published) Color(0xFF176345) else PearColors.SurfaceElevated,
                RoundedCornerShape(8.dp),
            )
            .padding(horizontal = 8.dp, vertical = 3.dp),
    )
}

@Composable
private fun ThemeRow(selected: SiteTheme, onSelect: (SiteTheme) -> Unit) {
    Row(Modifier.horizontalScroll(rememberScrollState())) {
        SiteTheme.presets.forEach { preset ->
            val active = preset.id == selected.id
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .padding(end = 8.dp)
                    .background(
                        if (active) PearColors.Accent else PearColors.Surface,
                        CircleShape,
                    )
                    .clickable { onSelect(preset) }
                    .padding(horizontal = 12.dp, vertical = 6.dp),
            ) {
                Spacer(
                    Modifier
                        .width(10.dp)
                        .height(10.dp)
                        .background(preset.swatch(), CircleShape),
                )
                Spacer(Modifier.width(6.dp))
                Text(
                    preset.label,
                    color = if (active) PearColors.Bg else PearColors.TextSecondary,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Medium,
                )
            }
        }
    }
}

private fun SiteTheme.swatch(): Color = when (id) {
    "default" -> PearColors.Accent
    "dark" -> Color.Black
    "warm" -> Color(0xFFF28C33)
    "ocean" -> PearColors.Link
    "forest" -> PearColors.Success
    else -> PearColors.Accent
}

private fun newBlock(type: BlockType): EditorBlock = when (type) {
    BlockType.Heading -> EditorBlock(type = type, text = "Heading")
    BlockType.List -> EditorBlock(type = type, items = listOf("Item 1"))
    BlockType.Link -> EditorBlock(type = type, text = "Link", href = "https://")
    else -> EditorBlock(type = type)
}

@Composable
private fun BlockEditorCard(
    block: EditorBlock,
    onChange: (EditorBlock) -> Unit,
    onMove: (Int) -> Unit,
    onRemove: () -> Unit,
) {
    Column(
        Modifier
            .fillMaxWidth()
            .background(PearColors.Surface.copy(alpha = 0.5f), RoundedCornerShape(10.dp))
            .padding(12.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                block.type.wire.uppercase(),
                color = PearColors.TextMuted,
                fontSize = 10.sp,
                fontWeight = FontWeight.SemiBold,
                letterSpacing = 1.sp,
                modifier = Modifier.weight(1f),
            )
            BlockControl("↑") { onMove(-1) }
            BlockControl("↓") { onMove(1) }
            BlockControl("x", color = PearColors.Error, onClick = onRemove)
        }
        Spacer(Modifier.height(6.dp))
        when (block.type) {
            BlockType.Heading -> EditorTextField(
                value = block.text,
                onValueChange = { onChange(block.copy(text = it)) },
                placeholder = "Heading",
                fontSize = 20.sp,
                fontWeight = FontWeight.Bold,
            )
            BlockType.Text, BlockType.Code, BlockType.Quote -> EditorTextField(
                value = block.text,
                onValueChange = { onChange(block.copy(text = it)) },
                placeholder = block.type.wire.replaceFirstChar { it.uppercase() },
                fontSize = if (block.type == BlockType.Code) 13.sp else 14.sp,
                fontFamily = if (block.type == BlockType.Code) FontFamily.Monospace else null,
                modifier = Modifier.heightIn(min = 80.dp),
            )
            BlockType.List -> {
                block.items.forEachIndexed { itemIndex, item ->
                    EditorTextField(
                        value = item,
                        onValueChange = { updated ->
                            onChange(block.copy(
                                items = block.items.mapIndexed { i, old -> if (i == itemIndex) updated else old },
                            ))
                        },
                        placeholder = "Item",
                    )
                    Spacer(Modifier.height(6.dp))
                }
                Text(
                    "+ item",
                    color = PearColors.Accent,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.SemiBold,
                    modifier = Modifier
                        .clickable { onChange(block.copy(items = block.items + "")) }
                        .padding(vertical = 4.dp),
                )
            }
            BlockType.Divider -> Spacer(
                Modifier
                    .fillMaxWidth()
                    .padding(vertical = 8.dp)
                    .height(1.dp)
                    .background(PearColors.Border),
            )
            BlockType.Link -> {
                EditorTextField(
                    value = block.text,
                    onValueChange = { onChange(block.copy(text = it)) },
                    placeholder = "Label",
                )
                Spacer(Modifier.height(6.dp))
                EditorTextField(
                    value = block.href,
                    onValueChange = { onChange(block.copy(href = it)) },
                    placeholder = "https://…",
                    fontFamily = FontFamily.Monospace,
                    noCapitalize = true,
                )
            }
            BlockType.Image -> {
                EditorTextField(
                    value = block.src,
                    onValueChange = { onChange(block.copy(src = it)) },
                    placeholder = "Image URL",
                    fontFamily = FontFamily.Monospace,
                    noCapitalize = true,
                )
                Spacer(Modifier.height(6.dp))
                EditorTextField(
                    value = block.alt,
                    onValueChange = { onChange(block.copy(alt = it)) },
                    placeholder = "Alt text",
                )
            }
        }
    }
}

@Composable
private fun BlockControl(
    label: String,
    color: Color = PearColors.TextSecondary,
    onClick: () -> Unit,
) {
    Text(
        label,
        color = color,
        fontSize = 14.sp,
        fontWeight = FontWeight.Bold,
        modifier = Modifier
            .clickable(onClick = onClick)
            .padding(horizontal = 8.dp, vertical = 2.dp),
    )
}

@Composable
private fun BlockToolbar(onAdd: (BlockType) -> Unit) {
    Surface(color = PearColors.Surface.copy(alpha = 0.9f)) {
        Column {
            Spacer(
                Modifier
                    .fillMaxWidth()
                    .height(0.5.dp)
                    .background(PearColors.Border),
            )
            Row(
                Modifier
                    .horizontalScroll(rememberScrollState())
                    .padding(horizontal = 16.dp, vertical = 10.dp),
            ) {
                BlockType.entries.forEach { type ->
                    Text(
                        type.toolbarLabel,
                        color = PearColors.TextPrimary,
                        fontSize = 14.sp,
                        fontWeight = FontWeight.Bold,
                        fontFamily = FontFamily.Monospace,
                        modifier = Modifier
                            .padding(end = 6.dp)
                            .background(PearColors.SurfaceElevated, RoundedCornerShape(8.dp))
                            .clickable { onAdd(type) }
                            .padding(horizontal = 12.dp, vertical = 8.dp),
                    )
                }
            }
        }
    }
}

@Composable
private fun EditorTextField(
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String,
    fontSize: androidx.compose.ui.unit.TextUnit = 14.sp,
    fontWeight: FontWeight? = null,
    fontFamily: FontFamily? = null,
    noCapitalize: Boolean = false,
    modifier: Modifier = Modifier,
) {
    TextField(
        value = value,
        onValueChange = onValueChange,
        placeholder = { Text(placeholder, color = PearColors.TextMuted, fontSize = fontSize) },
        textStyle = TextStyle(
            color = PearColors.TextPrimary,
            fontSize = fontSize,
            fontWeight = fontWeight,
            fontFamily = fontFamily,
        ),
        keyboardOptions = KeyboardOptions(
            capitalization = if (noCapitalize) KeyboardCapitalization.None else KeyboardCapitalization.Sentences,
            autoCorrectEnabled = false,
        ),
        colors = TextFieldDefaults.colors(
            focusedContainerColor = PearColors.Surface,
            unfocusedContainerColor = PearColors.Surface,
            focusedIndicatorColor = Color.Transparent,
            unfocusedIndicatorColor = Color.Transparent,
            cursorColor = PearColors.Accent,
            focusedTextColor = PearColors.TextPrimary,
            unfocusedTextColor = PearColors.TextPrimary,
        ),
        modifier = modifier.fillMaxWidth(),
    )
}
